server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/SnapshotInProgressAllocationDeciderTests.java

```diff
@@ -0,0 +1,240 @@
+            TestShardRouting.newShardRouting(shardId, nodeId, true, ShardRoutingState.STARTED),
+            null,
+            routingAllocation
+        );
+
+        assertEquals(Decision.Type.YES, decision.type());
+        assertEquals("no snapshots are currently running", decision.getExplanation());
+    }
+
+    public void testYesWhenNoShardSnapshot() {
+        final var routingAllocation = new RoutingAllocation(
+            new AllocationDeciders(List.of(decider)),
+            // snapshot in progress but not targetting this shard
+            makeClusterState(new ShardId(randomIdentifier(), randomUUID(), 0), randomFrom(SnapshotsInProgress.ShardState.values())),
+            ClusterInfo.EMPTY,
+            SnapshotShardSizeInfo.EMPTY,
+            randomNonNegativeLong()
+        );
+        routingAllocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+
+        final var decision = decider.canAllocate(
+            TestShardRouting.newShardRouting(shardId, nodeId, true, ShardRoutingState.STARTED),
+            null,
+            routingAllocation
+        );
+
+        assertEquals(Decision.Type.YES, decision.type());
+        assertEquals("the shard is not being snapshotted", decision.getExplanation());
+    }
+
+    public void testYesWhenShardSnapshotComplete() {
+        final var routingAllocation = new RoutingAllocation(
+            new AllocationDeciders(List.of(decider)),
+            // snapshot in progress but complete
+            makeClusterState(
+                shardId,
+                randomFrom(
+                    Arrays.stream(SnapshotsInProgress.ShardState.values()).filter(SnapshotsInProgress.ShardState::completed).toList()
+                )
+            ),
+            ClusterInfo.EMPTY,
+            SnapshotShardSizeInfo.EMPTY,
+            randomNonNegativeLong()
+        );
+        routingAllocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+
+        final var decision = decider.canAllocate(
+            TestShardRouting.newShardRouting(shardId, nodeId, true, ShardRoutingState.STARTED),
+            null,
+            routingAllocation
+        );
+
+        assertEquals(Decision.Type.YES, decision.type());
+        assertEquals("the shard is not being snapshotted", decision.getExplanation());
+    }
+
+    public void testYesWhenShardSnapshotOnDifferentNode() {
+        final var routingAllocation = new RoutingAllocation(
+            new AllocationDeciders(List.of(decider)),
+            makeClusterState(shardId, randomFrom(SnapshotsInProgress.ShardState.values())),
+            ClusterInfo.EMPTY,
+            SnapshotShardSizeInfo.EMPTY,
+            randomNonNegativeLong()
+        );
+        routingAllocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+
+        final var decision = decider.canAllocate(
+            // shard on a different node from the snapshot in progress one
+            TestShardRouting.newShardRouting(shardId, randomIdentifier(), true, ShardRoutingState.STARTED),
```

Comment:
Do we make snapshots of a primaries only or any shard copies?
Is this checking that a shard copy can move while we are making a snapshot of another one?

Reply:
Primaries only. But no, this is checking that the node IDs must match.

Reply:
I am trying to make sense of this scenario.
How could it happen that ShardSnapshotStatus and the ShardRouting contain different nodeIds for the primary shard that we are snapshotting?